compiler-access/src/main/scala/org/alephium/ralph/lsp/access/compiler/RalphCompilerAccess.scala

```diff
@@ -93,13 +95,93 @@ private object RalphCompilerAccess extends CompilerAccess {
+      // Ensure warnings with empty file information are logged so they can be resolved in Node.
+      warnings foreach {
+        warning =>
+          if (warning.sourceIndex.flatMap(_.fileURI).isEmpty)
+            logger.error(CompilerLogMessage.UnassignedWarning(warning))
+      }
+
       val statefulContracts =
-        statefulContractsAndIndex.map(_._1)
+        statefulContractsAndIndex.map(_._1).toArray
 
       val statefulScripts =
-        extendedContracts.genStatefulScripts()(options)
+        extendedContracts.genStatefulScripts()(options).toArray
+
+      val extractedContractWarnings =
+        extractWarnings(statefulContracts.map(CompiledCodeWrapper(_)))
 
-      Right((statefulContracts.toArray, statefulScripts.toArray, warnings.toArray))
+      val extractedScriptWarnings =
+        extractWarnings(statefulScripts.map(CompiledCodeWrapper(_)))
+
+      // Combine all warnings into a single collection so there is only one way to handle Warnings by presentation-compiler.
+      // This list contains all Warnings with SourceIndex and fileURI populated.
+      // If there are cases where SourceIndex or fileURI is None, they can be ignored by presentation-compiler,
+      // because they are logged.
+      val allWarnings =
+        extractedContractWarnings ++ extractedScriptWarnings ++ warnings
+
+      Right((statefulContracts, statefulScripts, allWarnings))
     } catch TryUtil.catchAllThrows(workspaceErrorURI)
 
+  private def extractWarnings(contracts: Array[CompiledCodeWrapper])(implicit logger: ClientLogger): Array[Warning] =
+    contracts flatMap extractWarnings
+
+  /**
+   * Ensures all [[Warning]]s are assigned a [[SourceIndex]] and file URI.
+   * Warnings that cannot be assigned are logged.
+   *
+   * Warnings returned by ralphc may or may not contain a [[SourceIndex]] or file URI.
+   *
+   * There multiple conditions for handling [[Warning]]s returned by ralphc
+   *  - When a warning contained in a [[CompiledContract]] or [[CompiledScript]]
+   *    has an empty [[SourceIndex]] or [[SourceIndex.fileURI]], the file information
+   *    for the contract or script must be applied to the [[Warning]].
+   *  - When a [[CompiledContract]] or [[CompiledScript]] contains a [[Warning]] for different file URI,
+   *    that [[Warning]] must be reported at the file information contained with the [[Warning]] and not
+   *    at the [[CompiledContract]]'s or [[CompiledScript]]'s file information.
+   *  - Since [[Warning.sourceIndex]] is of type [[Option]], [[None]] cases must be reported
+   *    so they can be resolved in Node.
+   *  - Since [[SourceIndex.fileURI]] is of type [[Option]], [[None]] cases must be reported
+   *    so they can be resolved in Node.
+   *
+   * @param contract The contract containing warnings.
+   * @param logger   Message logger
+   * @return Warnings with [[SourceIndex]] populated.
+   */
+  def extractWarnings(contract: CompiledCodeWrapper)(implicit logger: ClientLogger): Array[Warning] =
```

Comment:
[nitpick] in the first match your `if else` as no brackets, while the second one as brackets